### ISSUE TITLE
DuplicateKeyError with counterblockd

### ISSUE THREAD

**Issue by toysrtommy** [CounterpartyXCP/counterblock#62]
Counterblockd throws duplicatekeyerror, and can't take any request (always complain Server is not caught up. Please try again later.)

Here's the full log:
xcp@ip-172-31-34-102:~/counterblockd$ python counterblockd.py --testnet --counterpartyd-rpc-user rpc --counterpartyd-rpc-password ipYcMLljCmtrDI

2014-10-03 22:09:52,113 :: INFO :: counterblock Version 1.5.0 starting ...
2014-10-03 22:09:52,113 :: WARNING :: Counterwallet config does not exist at '/home/xcp/counterwallet/counterwallet.conf.json'
2014-10-03 22:09:52,114 :: INFO :: Checking/updating GeoIP.dat ...
2014-10-03 22:09:52,114 :: INFO :: GeoIP.dat OK
2014-10-03 22:09:52,114 :: INFO :: Connecting to mongoDB backend ...
2014-10-03 22:09:52,137 :: INFO :: Starting up socket.io server (block event feed)...
2014-10-03 22:09:52,138 :: INFO :: Starting up socket.io server (chat feed)...
2014-10-03 22:09:52,138 :: INFO :: Starting up counterpartyd block feed poller...
2014-10-03 22:09:52,138 :: INFO :: Starting up RPC API handler...
2014-10-03 22:09:52,146 :: WARNING :: Pruning to block 294632 ...
Traceback (most recent call last):
  File "/usr/lib/python2.7/gevent/greenlet.py", line 327, in run
    result = self._run(*self.args, **self.kwargs)
  File "/home/xcp/counterblockd/lib/events.py", line 27, in check_blockchain_service
    raise Exception('Could not connect to blockchain service: %s' % e)
Exception: Could not connect to blockchain service: Insight reports error: Safe mode: Warning: The network does not appear to fully agree! Some miners appear to be experiencing issues.
<Greenlet at 0x7fbf4dd37cd0: check_blockchain_service> failed with Exception

http://localhost:14000/api/
2014-10-03 22:09:52,156 :: INFO :: Revised wallet statistics for partial day 2014-6-13: {'referers': {u'': 1}, 'network': 'testnet', 'new_count': 2}
http://localhost:14000/api/
http://localhost:14000/api/
2014-10-03 22:09:52,166 :: INFO :: Received message 44246: {u'category': u'debits', u'message_index': 44246, u'timestamp': 1412374161, u'block_index': 294633, u'command': u'insert', u'bindings': u'{"action": null, "address": "msLQdj49Yfg6AEMo3wsnFDTTKLyNZbzk5j", "asset": "GLDUSAEAGLOZ", "block_index": 294633, "event": "99dfd594b4c339de4fe157ff269762d38c6d87480208a29e1005b7721d3add7b", "quantity": 1}'} ...
2014-10-03 22:09:52,167 :: INFO :: Procesed debit bal change from tx 44246 :: {'new_balance': -1, 'block_index': 294633, 'quantity_normalized': -1, 'block_time': datetime.datetime(2014, 10, 3, 22, 0, 45), 'asset': u'GLDUSAEAGLOZ', 'new_balance_normalized': -1, 'address': u'msLQdj49Yfg6AEMo3wsnFDTTKLyNZbzk5j', '_id': ObjectId('542f1eb076ac144e73c02446'), 'quantity': -1}
2014-10-03 22:09:52,167 :: INFO :: Received message 44247: {u'category': u'credits', u'message_index': 44247, u'timestamp': 1412374161, u'block_index': 294633, u'command': u'insert', u'bindings': u'{"action": null, "address": "moNEdvs1YctnA5MNUm9fyRDbzLnE1DmvaA", "asset": "GLDUSAEAGLOZ", "block_index": 294633, "event": "99dfd594b4c339de4fe157ff269762d38c6d87480208a29e1005b7721d3add7b", "quantity": 1}'} ...
2014-10-03 22:09:52,169 :: INFO :: Procesed credit bal change from tx 44247 :: {'new_balance': 98, 'block_index': 294633, 'quantity_normalized': 1, 'block_time': datetime.datetime(2014, 10, 3, 22, 0, 45), 'asset': u'GLDUSAEAGLOZ', 'new_balance_normalized': 98, 'address': u'moNEdvs1YctnA5MNUm9fyRDbzLnE1DmvaA', '_id': ObjectId('542f1eb076ac144e73c02447'), 'quantity': 1}
2014-10-03 22:09:52,169 :: INFO :: Received message 44248: {u'category': u'sends', u'message_index': 44248, u'timestamp': 1412374161, u'block_index': 294633, u'command': u'insert', u'bindings': u'{"asset": "GLDUSAEAGLOZ", "block_index": 294633, "destination": "moNEdvs1YctnA5MNUm9fyRDbzLnE1DmvaA", "quantity": 1, "source": "msLQdj49Yfg6AEMo3wsnFDTTKLyNZbzk5j", "status": "valid", "tx_hash": "99dfd594b4c339de4fe157ff269762d38c6d87480208a29e1005b7721d3add7b", "tx_index": 226091}'} ...
Traceback (most recent call last):
  File "/usr/lib/python2.7/gevent/greenlet.py", line 327, in run
    result = self._run(*self.args, **self.kwargs)
  File "/home/xcp/counterblockd/lib/blockfeed.py", line 306, in process_cpd_blockfeed
    'category': msg['category']
  File "/usr/local/lib/python2.7/dist-packages/pymongo/collection.py", line 410, in insert
    _check_write_command_response(results)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/helpers.py", line 198, in _check_write_command_response
    raise DuplicateKeyError(error.get("errmsg"), 11000, error)
DuplicateKeyError: insertDocument :: caused by :: 11000 E11000 duplicate key error index: counterblockd_testnet.transaction_stats.$message_index_1  dup key: { : 44248 }
<Greenlet at 0x7fbf4dd37e10: process_cpd_blockfeed(<zmq.green.core._Socket object at 0x7fbf4dcc1530>)> failed with DuplicateKeyError

fcTraceback (most recent call last):
  File "/usr/lib/python2.7/gevent/greenlet.py", line 327, in run
    result = self._run(*self.args, **self.kwargs)
  File "/home/xcp/counterblockd/lib/events.py", line 27, in check_blockchain_service
    raise Exception('Could not connect to blockchain service: %s' % e)
Exception: Could not connect to blockchain service: Insight reports error: Safe mode: Warning: The network does not appear to fully agree! Some miners appear to be experiencing issues.
<Greenlet at 0x7fbf4dce5050: check_blockchain_service> failed with Exception

**Comment by rippler** [comment]
@robby-dermody I suppose this can be closed as it was fixed in 9.49.3

**Comment by curtislacy** [comment]
I think we saw this today, with 9.49.3.  Counterblockd was consistently crashing on block 342788, and running it manually yielded this error:

```
INFO:messages: Procesed credit bal change from tx 594975 :: {'new_balance': 57867771088, 'block_index': 342788, 'quantity_normalized': 578.67771088, 'block_time': datetime.datetime(2015, 2, 10, 1, 58, 27), 'asset': u'GEMZ', 'new_balance_normalized': 578.67771088, 'address': u'1N1j8KJd93QqgLVWtAiuQAMU1tqhxUAff8', '_id': ObjectId('54de6893be4f766bd231f816'), 'quantity': 57867771088}
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/gevent/greenlet.py", line 340, in run
    result = self._run(*self.args, **self.kwargs)
  File "/home/bitcoin/counterblockd/lib/blockfeed.py", line 244, in process_cpd_blockfeed
    parse_block(block_data)
  File "/home/bitcoin/counterblockd/lib/blockfeed.py", line 113, in parse_block
    cmd = parse_message(msg)
  File "/home/bitcoin/counterblockd/lib/blockfeed.py", line 98, in parse_message
    cmd = function['function'](msg, msg_data) or None
  File "/home/bitcoin/counterblockd/lib/processor/messages.py", line 49, in parse_insert
    'category': msg['category']
  File "/usr/local/lib/python2.7/dist-packages/pymongo/collection.py", line 415, in insert
    self.uuid_subtype, client)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/mongo_client.py", line 1124, in _send_message
    rv = self.__check_response_to_last_error(response, command)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/mongo_client.py", line 1065, in __check_response_to_last_error
    raise DuplicateKeyError(details["err"], code, result)
DuplicateKeyError: E11000 duplicate key error index: counterblockd.transaction_stats.$message_index_1  dup key: { : 594976 }
<Greenlet at 0x7fd849e99eb0: process_cpd_blockfeed(<zmq.green.core._Socket object at 0x7fd848d54258>)> failed with DuplicateKeyError
```

Reparse is running now, waiting to see if that clears it up....
